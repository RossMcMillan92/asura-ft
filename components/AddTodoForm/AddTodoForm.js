import Button from '../Button/Button'
import PropTypes from 'prop-types'
import React from 'react'
import TextInput from '../TextInput/TextInput'
import { trim } from 'lodash/fp'

const AddTodoForm = ({ onSubmit }) => {
  const [newTodo, updateNewTodo] = React.useState('')

  return (
    <form
      className="u-flex-row u-width-100%"
      onSubmit={e => {
        e.preventDefault()
        const trimmedTodo = trim(newTodo)
        if (trimmedTodo) onSubmit(trimmedTodo)
        updateNewTodo('')
      }}
    >
      <TextInput
        autoComplete="off"
        id="todo-entry"
        name="todo-entry"
        className="u-flex-grow-1 u-margin-right-20"
        label="Enter a task..."
        placeholder="Enter a task..."
        aria-invalid={newTodo.length === 0}
        onChange={e => updateNewTodo(e.target.value)}
        value={newTodo}
        required
      />
      <Button>Add</Button>
    </form>
  )
}

AddTodoForm.propTypes = {
  onSubmit: PropTypes.func.isRequired
}

export default AddTodoForm
