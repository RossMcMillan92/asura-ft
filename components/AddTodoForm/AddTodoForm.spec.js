import { mount, shallow } from 'enzyme'

import AddTodoForm from './AddTodoForm'
import React from 'react'
import { act } from 'react-dom/test-utils'

const props = {
  onSubmit: jest.fn()
}
const getWrapper = additionalProps =>
  mount(<AddTodoForm {...props} {...additionalProps} />)
const getShallowWrapper = additionalProps =>
  shallow(<AddTodoForm {...props} {...additionalProps} />)

describe('AddTodoForm', () => {
  it('matches the snapshot', () => {
    const wrapper = getShallowWrapper()
    expect(wrapper).toMatchSnapshot()
  })

  describe('when submitted', () => {
    describe('when the text field is empty', () => {
      it('does not call the on submit prop', () => {
        const wrapper = getWrapper()
        act(() => {
          wrapper.find('form').simulate('submit')
        })
        expect(props.onSubmit).not.toBeCalled()
      })
    })

    describe('when the text field is filled', () => {
      const wrapper = getWrapper()
      const value = '   To do task... '

      beforeAll(() => {
        act(() => {
          wrapper.find('input').simulate('change', { target: { value } })
        })
        wrapper.update()

        // Ensure the value is in the text entry before later checking it's cleared
        expect(wrapper.find('input').prop('value')).toBe(value)

        act(() => {
          wrapper.find('form').simulate('submit')
        })
        wrapper.update()
      })

      it('calls the onSubmit prop, passing the trimmed value', () => {
        expect(props.onSubmit).toBeCalledWith(value.trim())
      })

      it('clears the text field', () => {
        expect(wrapper.find('input').prop('value')).toBe('')
      })
    })
  })
})
