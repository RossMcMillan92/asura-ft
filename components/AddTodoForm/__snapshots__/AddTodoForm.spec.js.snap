// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodoForm matches the snapshot 1`] = `
<form
  className="u-flex-row u-width-100%"
  onSubmit={[Function]}
>
  <TextInput
    aria-invalid={true}
    autoComplete="off"
    className="u-flex-grow-1 u-margin-right-20"
    id="todo-entry"
    label="Enter a task..."
    name="todo-entry"
    onChange={[Function]}
    placeholder="Enter a task..."
    required={true}
    value=""
  />
  <Button>
    Add
  </Button>
</form>
`;
